modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/service/impl/DDMStructureLocalServiceImpl.java

```diff
@@ -1510,18 +1510,28 @@ public DDMStructure updateStructure(
 	@Override
 	public DDMStructure updateStructure(
 			long userId, long structureId, long parentStructureId,
-			Map<Locale, String> nameMap, Map<Locale, String> descriptionMap,
-			String definition, ServiceContext serviceContext)
+			String structureKey, Map<Locale, String> nameMap,
+			Map<Locale, String> descriptionMap, String definition,
+			ServiceContext serviceContext)
 		throws PortalException {
 
 		DDMStructure structure = ddmStructurePersistence.findByPrimaryKey(
 			structureId);
 
 		User user = _userLocalService.getUser(userId);
 
+		_validate(
+			structure.getGroupId(), structure.getParentStructureId(),
+			structure.getClassNameId(), structureKey, nameMap,
+			structure.getDDMForm());
+
 		structure.setUserId(userId);
 		structure.setParentStructureId(parentStructureId);
 
+		if (Validator.isNotNull(structureKey)) {
+			structure.setStructureKey(structureKey);
```

Comment:
@kristofloneck can we make it similar to the [addStructure](https://github.com/liferay/liferay-portal/blob/570b5a7c2db64db5c62b4494c3cc29d12cbc17df/modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/service/impl/DDMStructureLocalServiceImpl.java#L254-L270) logic? Something like:

```
User user = _userLocalService.getUser(userId);

structure.setUserId(userId);
structure.setParentStructureId(parentStructureId);

if (Validator.isNotNull(structureKey)) {
	structureKey = StringUtil.toUpperCase(structureKey.trim());

	_validateStructureKey(
		structureId, structure.getGroupId(), structure.getClassNameId(),
		structureKey);

	structure.setStructureKey(structureKey);
}
```

Do we need to call the existing `_validate`? If I understand correctly, we only need to validate the `structureKey`, that's why I created the `_validateStructureKey` method

```
private void _validateStructureKey(
		long structureId, long groupId, long classNameId,
		String structureKey)
	throws PortalException {

	DDMStructure structure = ddmStructurePersistence.fetchByG_C_S(
		groupId, classNameId, structureKey);

	if ((structure == null) ||
		(structure.getStructureId() == structureId)) {

		return;
	}

	StructureDuplicateStructureKeyException
		structureDuplicateStructureKeyException =
			new StructureDuplicateStructureKeyException();

	structureDuplicateStructureKeyException.setStructureKey(
		structure.getStructureKey());

	throw structureDuplicateStructureKeyException;
}
```

Reply:
Made the changes requested by you